pymeasure/instruments/oxfordinstruments/ips120_10.py

```diff
@@ -0,0 +1,330 @@
+
+    def enable_control(self):
+        """ Method that enables active control of the IPS.
+        Sets control to remote and turns off the clamp. """
+        self.control_mode = "RU"
+
+        # Turn off clamping if still clamping
+        if self.activity == "clamp":
+            self.activity = "hold"
+
+        # Turn on switch-heater if field at zero
+        if self.field == 0:
+            self.switch_heater_status = 1
+
+    def disable_control(self):
+        if not self.field == 0:
+            raise Exception("IPS not at 0T, not disabling the supply")
+
+        self.switch_heater_status = 0
+        self.activity = "clamp"
+
+    def enable_persistent_mode(self):
+        """ Methods that enables the persistent magnetic field mode. """
+        # Check if system idle
+        if not self.sweep_status == "at rest":
+            raise Exception("Magnet not at rest")
+
+        switch_status = self.switch_heater_status
+        if switch_status in [0, 2]:
+            return  # Magnet already in persistent mode
+        elif switch_status == 1:
+            self.activity = "hold"
+            self.switch_heater_status = 0
+            log.info("IPS: Wait for 20s for switch heater")
+            sleep(self._SWITCH_HEATER_DELAY)
+            self.activity = "to zero"
+            self.wait_for_idle()
+        else:
+            raise Exception("Switch status returned %d" % switch_status)
+
+    def disable_persistent_mode(self):
+        """ Methods that disables the persistent magnetic field mode. """
+        # Check if system idle
+        if not self.sweep_status == "at rest":
+            raise Exception("Magnet not at rest")
+
+        # Check if the setpoint equals the persistent field
+        if not self.field == self.field_setpoint:
+            log.warning("IPS: field setpoint and persistent field not identical; "
+                        "setting the setpoint to the persistent field.")
+            self.field_setpoint = self.field
+
+        switch_status = self.switch_heater_status
+        if switch_status == 1:
+            return  # Magnet already in demand mode or at 0 field
+        elif switch_status in [0, 2]:
+            self.activity = "to setpoint"
+            self.wait_for_idle()
+            self.activity = "hold"
+            self.switch_heater_status = 1
+            log.info("IPS: Wait for 20s for switch heater")
+            sleep(self._SWITCH_HEATER_DELAY)
+        else:
+            raise Exception("Switch status returned %d" % switch_status)
+
+    def wait_for_idle(self, max_errors=10, delay=1):
+        error_ct = 0
+        status = None
+        while True:
+            sleep(delay)
+
+            try:
+                status = self.sweep_status
+            except Exception as e:
+                log.error("IPS: Issue with getting status (#%d): %s" % (error_ct, e))
+                error_ct = 0
+
+            if status == "at rest":
+                break
+
+            if error_ct > max_errors:
```

Comment:
you seem to never increment `error_ct`, so can that ever become true (except for negative `max_errors`?) :-)

Reply:
Ah nice catch, thanks